tests/import_specifications/test_file_parser.py

```diff
@@ -42,39 +115,69 @@ def test_Error_init_w_OTHER_success():
 
     assert e.error == ErrorType.OTHER
     assert e.message == "foo"
-    assert e.source == None
+    assert e.source_1 is None
+    assert e.source_2 is None
 
     # all
-    e = Error(ErrorType.OTHER, message="foo", source=SpecificationSource("whoo"))
+    e = Error(ErrorType.OTHER, message="foo", source_1=spcsrc("wooo"))
 
     assert e.error == ErrorType.OTHER
     assert e.message == "foo"
-    assert e.source == SpecificationSource("whoo")
+    assert e.source_1 == spcsrc("wooo")
+    assert e.source_2 is None
 
 
 def test_Error_init_fail():
-    error_init_fail(ErrorType.FILE_NOT_FOUND, None, None, ValueError(
-        "source is required for a FILE_NOT_FOUND error"))
-    err = "source and message are required for a PARSE_FAIL error"
-    error_init_fail(ErrorType.PARSE_FAIL, None, SpecificationSource("foo"), ValueError(err))
-    error_init_fail(ErrorType.PARSE_FAIL, "msg", None, ValueError(err))
-    error_init_fail(ErrorType.OTHER, None, None, ValueError(
+    error_init_fail(None, None, None, None, ValueError("error is required"))
+    error_init_fail(ErrorType.FILE_NOT_FOUND, None, None, None, ValueError(
+        "source_1 is required for a FILE_NOT_FOUND error"))
+    err = "message and source_1 are required for a PARSE_FAIL error"
+    error_init_fail(ErrorType.PARSE_FAIL, None, spcsrc("wooo"), None, ValueError(err))
+    error_init_fail(ErrorType.PARSE_FAIL, "msg", None, None, ValueError(err))
+    ms = ErrorType.MULTIPLE_SPECIFICATIONS_FOR_DATA_TYPE
+    err = ("message, source_1, and source_2 are required for a "
+        + "MULTIPLE_SPECIFICATIONS_FOR_DATA_TYPE error")
+    error_init_fail(ms, None, None, None, ValueError(err))
+    error_init_fail(ms, None, spcsrc("foo"), spcsrc("bar"), ValueError(err))
+    error_init_fail(ms, "msg", None, spcsrc("bar"), ValueError(err))
+    error_init_fail(ms, "msg", spcsrc("foo"), None, ValueError(err))
+    error_init_fail(ErrorType.OTHER, None, None, None, ValueError(
         "message is required for a OTHER error"))
 
 
-def error_init_fail(errortype, message, source, expected):
+def error_init_fail(errortype, message, source_1, source_2, expected):
+    with raises(Exception) as got:
+        Error(errortype, message, source_1, source_2)
+    assert_exception_correct(got.value, expected)
+
+
+def test_ParseResult_init_success():
+    pr = ParseResult(spcsrc("bar"), (frozendict({"foo": "bar"}),))
+
+    assert pr.source == spcsrc("bar")
+    assert pr.result == (frozendict({"foo": "bar"}),)
+
+
+def test_ParseResult_init_fail():
+    parseResult_init_fail(None, None, ValueError("source is required"))
+    parseResult_init_fail(None, (frozendict({"foo": "bar"}),), ValueError("source is required"))
+    parseResult_init_fail(spcsrc("foo"), None, ValueError("result is required"))
+
+
+def parseResult_init_fail(source, result, expected):
```

Comment:
Could add typehints here

Reply:
This one seems to be an outlier as well - there are many functions that are very similar to this one that don't have typehints. It'd be helpful if you could explain why this one in particular needs typehints as I don't really understand what the difference is

Reply:
I think we could add typehints to all of them. I'm in favor of being able to tell more about what the function needs and does just from the signature. Source and result and expected don't really mean much to me. I have to look inside the function and then see that `source` and `result` are arguments to ParseResult. I haven't memorized the arguments to ParseResult, so  then I need to look into ParseResult, and see that its actually a SpecificationSource and `tuple[frozendict[str, PRIMITIVE_TYPE]` . Then expected is actually an exception that goes into assert_exception_correct.

If it was something like
`def parseResult_init_fail(source: SpecificationSource, result: Result, expected: Exception):`

I could tell these things apart faster and easier, and if I forget for the next time, it's faster to go through the function knowing what's going on. Having a docstring about this test helper and what it does also would allow me to have an idea of what's going on before reading the code.

Reply:
I've always viewed typehints in test code as overkill, but you've convinced me - typehints everywhere. I'll take care of it in a future PR though.

That also means I'll be expecting test typehints in your code as well :P